README.md

```diff
@@ -23,6 +23,72 @@ Trying to make sense of the Siemens LOGO sort-of built-in REST API. Maybe, somed
+
+Tokens seem to have unlimited lifetime. Maybe they get lost when rebooting the LOGO. The "keep me logged in" checkbox only seems to tell the browser to store the Security-Hint. Checking or Unchecking the box has no effect on the messages sent and received from the LOGO.
+
+1. pick four random numbers between 0 and 2^32-1 (4294967295) and store them in variables `a1`, `a2`, `b1`, `b2`.
+
+2. POST-Request to http://logo-ip/AJAX with Header "Security-Hint: p" and body "UAMCHAL:3,4,`a1`,`a2`,`b1`,`b2`". Example: "UAMCHAL:3,4,1234567890,2345678901,3456789012,456789012"
+
+3. you'll receive a comma-separated string with three sections as reply: status-code, `Login Security-Hint`, and the `serverChallenge`. For example: "700,ABCDEF1234567890ABCDEF1234567890,567890123". Split on "," and check that the split-result has 3 parts, that the first part is int(700) and the third part is also an integer. Parse the `serverChallenge`as uint32 for all further steps.
+
+4. create the PW token base string by concating the password, a + sign, and the serverChallenge: 
+`pwToken = 'password' + '+' + serverChallenge`
+If the password has non-ansi characters, they need to be UTF8-encoded in the string.
+
+5. truncate the `pwToken` to max 32 chars. Note: this will do nothing unless your password is longer than 21 chars. I have not tested the login process with such a long password.
+
+6. get the CRC32 of the `pwToken`and xor it with the server challenge. This will be the pwToken for the server: `loginPWToken = crc32(pwToken) ^ serverChallenge`
+Use the serverChallenge in its uint32 form here.
+
+7. create the loginServerChallenge by xor-ing all four keys and the serverChallenge: `loginServerChallenge = a1 ^ a2 ^ b1 ^ b2 ^ serverChallenge`
+
+8. POST-Request to http://logo-ip/AJAX with Header "Security-Header: `Login Security-Hint`" and body "UAMLOGIN:Web User,`loginPWToken`,`loginServerChallenge`". Example: "UAMLOGIN:Web User,1345678901,1456789012".
+
+Here is a working example in PHP with gmp and curl extension:
+
+```php
+$pw = 'password';
+
+$a1 = gmp_random_range(0, "4294967296");
+$a2 = gmp_random_range(0, "4294967296");
+$b1 = gmp_random_range(0, "4294967296");
+$b2 = gmp_random_range(0, "4294967296");
+
+$ch = curl_init();
+curl_setopt($ch, CURLOPT_URL, 'http://[logo-ip]/AJAX');
+curl_setopt($ch, CURLOPT_HTTPHEADER, array('Security-Hint: p'));
+curl_setopt($ch, CURLOPT_POST, true);
+curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+curl_setopt($ch, CURLOPT_POSTFIELDS, "UAMCHAL:3,4,$a1,$a2,$b1,$b2");
+
+$ret = curl_exec($ch); //700,ABCDEF1234567890ABCDEF1234567890,1234567890
+$ret = explode(",", $ret);
+if (count($ret) == 3 && $ret[0] == "700") {
+  $pwToken = substr($pw . '+' . $ret[2], 0, 32);
+  $iPWToken = gmp_xor(crc32($pwToken), $ret[2]);
+  $iServerChallenge = gmp_xor(gmp_xor(gmp_xor(gmp_xor($a1, $a2), $b1), $b2), $challenge);
```

Comment:
$iServerChallenge = gmp_xor(gmp_xor(gmp_xor(gmp_xor($a1, $a2), $b1), $b2), $ret[2]);

Reply:
sorry, this fix is of course correct. It happened when I've tidied up my script for the pull request. I'll fix this